sql: TestStatementTimeoutForSchemaChangeCommit failed

sql.TestStatementTimeoutForSchemaChangeCommit [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21263154?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21263154?buildTab=artifacts#/) on release-26.1 @ [a882b0265976a1c77beee4a9d750c1322f65eac8](https://github.com/cockroachdb/cockroach/commits/a882b0265976a1c77beee4a9d750c1322f65eac8):


```
=== RUN   TestStatementTimeoutForSchemaChangeCommit
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/f4d1e5a362e4384d18c9c26e2defb957/logTestStatementTimeoutForSchemaChangeCommit3707551216
    test_log_scope.go:82: use -show-logs to present logs inline
    run_control_test.go:1165: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f4d1e5a362e4384d18c9c26e2defb957/logTestStatementTimeoutForSchemaChangeCommit3707551216
--- FAIL: TestStatementTimeoutForSchemaChangeCommit (2.99s)
=== RUN   TestStatementTimeoutForSchemaChangeCommit/implicitTxn=true
    test_server_shim.go:182: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:668: DRPC is enabled (via override by TestingGlobalDRPCOption)
    run_control_test.go:1148: 
        	Error Trace:	pkg/sql/run_control_test.go:1148
        	Error:      	Error "pq: job 1158967129871941633 could not be loaded: txn exec: context canceled" does not contain "query execution canceled due to statement timeout"
        	Test:       	TestStatementTimeoutForSchemaChangeCommit/implicitTxn=true
--- FAIL: TestStatementTimeoutForSchemaChangeCommit/implicitTxn=true (1.88s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatementTimeoutForSchemaChangeCommit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-61701

--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -4384,8 +4384,10 @@ func (ex *connExecutor) runWithStatementTimeout(
 			var cancelFn context.CancelFunc
 			waitCtx, cancelFn = context.WithCancel(waitCtx)
 			queryTimeTicker := time.AfterFunc(ex.sessionData().StmtTimeout-timePassed, func() {
-				cancelFn()
+				// Before the cancellation make sure that the query
+				// is marked as timed out.
 				queryTimedout.Store(true)
+				cancelFn()
 			})
 			defer cancelFn()
 			defer queryTimeTicker.Stop()

